**Threshold Optimization techniques to consider:**
1. ROC Curve largest distance from baseline (Better for balanced)
2. Precision - Recall curve intersection (Better for imbalanced)
3. G mean
4. J statistic value
5. Optimized F1 measure

# Project Overview

## Background and Context

AllLife Bank is a US bank that has a growing customer base. The majority of these customers are liability customers (depositors) with varying sizes of deposits. The number of customers who are also borrowers (asset customers) is quite small, and the bank is interested in expanding this base rapidly to bring in more loan business and in the process, earn more through the interest on loans. In particular, the management wants to explore ways of converting its liability customers to personal loan customers (while retaining them as depositors).

A campaign that the bank ran last year for liability customers showed a healthy conversion rate of over 9% success. This has encouraged the retail marketing department to devise campaigns with better target marketing to increase the success ratio.

You as a Data scientist at AllLife bank have to build a model that will help the marketing department to identify the potential customers who have a higher probability of purchasing the loan.

## Objective

1. To predict whether a liability customer will buy a personal loan or not.
2. Which variables are most significant.
3. Which segment of customers should be targeted more.

## Data Dictionary
* ID: Customer ID
* Age: Customer’s age in completed years
* Experience: #years of professional experience
* Income: Annual income of the customer (in thousand dollars)
* ZIP Code: Home Address ZIP code.
* Family: the Family size of the customer
* CCAvg: Average spending on credit cards per month (in thousand dollars)
* Education: Education Level. 1: Undergrad; 2: Graduate;3: Advanced/Professional
* Mortgage: Value of house mortgage if any. (in thousand dollars)
* Personal_Loan: Did this customer accept the personal loan offered in the last campaign?
* Securities_Account: Does the customer have securities account with the bank?
* CD_Account: Does the customer have a certificate of deposit (CD) account with the bank?
* Online: Do customers use internet banking facilities?
* CreditCard: Does the customer use a credit card issued by any other Bank (excluding All life Bank)?

# Data Investigation

## Install and load libraries as needed

In [ ]:
!pip install uszipcode

In [33]:
!pip install -U googlemaps

  Created wheel for googlemaps: filename=googlemaps-4.6.0-py3-none-any.whl size=38552 sha256=5aa71a04d4809f405da49015293928306dd9994eac11ba2d98b46c35c14427d0
  Stored in directory: /Users/Shyamal/Library/Caches/pip/wheels/b6/23/aa/a130b53682531f8382dd06f2000001225bea7b56a99c516f87
Successfully built googlemaps


<IPython.core.display.Javascript object>

In [1]:
# this will help in making the Python code more structured automatically (good coding practice)
%load_ext nb_black

import warnings

warnings.filterwarnings("ignore")

# Libraries to help with reading and manipulating data

import pandas as pd
import numpy as np

# Library to split data
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer

# Libraries to help with data visualization
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline 

# Selecting pandas plotting backend 
pd.options.plotting.backend = "hvplot"

# Removes the limit for the number of displayed columns
pd.set_option("display.max_columns", None)
# Sets the limit for the number of displayed rows
pd.set_option("display.max_rows", 200)


# To build model for prediction
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree
from sklearn.model_selection import GridSearchCV

# To work with categorical features that have order
from sklearn.preprocessing import OrdinalEncoder

# Plot ROC Curve
from sklearn.metrics import plot_roc_curve

# To get diferent metric scores
from sklearn.metrics import (
    f1_score,
    accuracy_score,
    recall_score,
    precision_score,
    confusion_matrix,
    roc_auc_score,
    plot_confusion_matrix,
    precision_recall_curve,
    roc_curve,
)

<IPython.core.display.Javascript object>

In [2]:
%matplotlib inline 

<IPython.core.display.Javascript object>

In [3]:
#from google.colab import drive
#drive.mount('/content/drive')

<IPython.core.display.Javascript object>

## Load data and review

In [4]:
#path = ('/content/drive/MyDrive/Colab Notebooks/Project_3/Loan_Modelling.csv')
path = 'Loan_Modelling.csv'
data = pd.read_csv(path)
df = data.copy()

<IPython.core.display.Javascript object>

In [5]:
# Dataframe shape
print(f'Dataframe has {df.shape[0]} rows and {df.shape[1]} columns.')

Dataframe has 5000 rows and 14 columns.


<IPython.core.display.Javascript object>

In [6]:
# Type of data in dataframe
df.dtypes.value_counts()

int64      13
float64     1
dtype: int64

<IPython.core.display.Javascript object>

* Our dataset only contains numerical features where all but one are intigers. Looking at the data dictionary there might be categorical features that are already encoded. We can check this out below

In [7]:
# Sampling our dataset to see what we're dealing with
df.sample(20)

ID  Age  Experience  Income  ZIPCode  Family  CCAvg  Education  \
287    288   37          12      62    92110       3   0.70          2   
4313  4314   52          28      79    94596       1   2.70          2   
4202  4203   35           9      82    95064       3   0.90          2   
1544  1545   39          15      24    92123       1   1.00          1   
4176  4177   44          18      75    95131       1   0.70          3   
1465  1466   45          19      60    91911       1   0.70          3   
2545  2546   25          -1      39    94720       3   2.40          2   
3662  3663   35           9     164    94305       2   0.00          1   
1734  1735   35          10      79    94720       4   2.10          3   
1904  1905   38          14      91    95060       2   0.00          1   
2877  2878   58          32      74    95817       2   2.30          3   
4221  4222   48          22      83    90028       2   0.40          3   
4808  4809   42          16      32    90033       3   0.50          2   
734    735   66          42      53    92182       2   1.10          1   
1189  1190   42          17     115    92717       2   0.40          1   
1188  1189   45          19      58    94305       2   0.40          3   
3673  3674   34           9      65    95134       4   1.30          1   
152    153   57          32      24    93117       1   1.30          1   
1195  1196   32           7     123    92407       2   2.90          2   
1389  1390   45          15      20    94107       1   0.75          3   

      Mortgage  Personal_Loan  Securities_Account  CD_Account  Online  \
287          0              0                   0           0       0   
4313         0              0                   0           0       1   
4202         0              0                   0           0       1   
1544       116              0                   0           0       1   
4176         0              0                   0           0       0   
1465       159              0                   0           0       1   
2545         0              0                   0           0       1   
3662       500              0                   0           0       0   
1734       182              0                   0           0       1   
1904         0              0                   0           0       1   
2877         0              0                   0           0       0   
4221       248              0                   0           0       1   
4808         0              0                   0           0       0   
734          0              0                   0           0       1   
1189         0              0                   0           0       1   
1188         0              0                   1           0       0   
3673         0              0                   0           0       0   
152          0              0                   0           0       1   
1195         0              1                   0           0       1   
1389         0              0                   0           0       1   

      CreditCard  
287            1  
4313           1  
4202           0  
1544           1  
4176           1  
1465           0  
2545           0  
3662           0  
1734           0  
1904           0  
2877           1  
4221           0  
4808           1  
734            1  
1189           0  
1188           1  
3673           1  
152            1  
1195           0  
1389           0

<IPython.core.display.Javascript object>

* Dtypes match up with what we see from the sampled list
* Zipcode is numerical but also a categorical feature
* Last 5 columns look like they're already encoded but we will verify this later 
* Education is also already encoded but more so a categorical feature

In [8]:
# Print numerical and categorical features in data set
cat_cols = ['ZIPCode', 'Education'] 
num_cols = ['ID', 'Age', 'Experience', 'Income', 'Family', 'CCAvg',
            'Mortgage', 'Personal_Loan', 'Securities_Account',
            'CD_Account', 'Online', 'CreditCard']
print('Categorical columns: \n',cat_cols)
print('Numeric columns: \n', num_cols)

Categorical columns: 
 ['ZIPCode', 'Education']
Numeric columns: 
 ['ID', 'Age', 'Experience', 'Income', 'Family', 'CCAvg', 'Mortgage', 'Personal_Loan', 'Securities_Account', 'CD_Account', 'Online', 'CreditCard']


<IPython.core.display.Javascript object>

In [9]:
# Check for duplicates in our data
df.duplicated().sum()

0

<IPython.core.display.Javascript object>

In [10]:
# Check for null values 
df.isnull().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIPCode               0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal_Loan         0
Securities_Account    0
CD_Account            0
Online                0
CreditCard            0
dtype: int64

<IPython.core.display.Javascript object>

- Looks like our dataset doens't contain any duplicates or null values

## Structure of "Categorical" Features

In [11]:
# Printing off the first 5 rows for a quick recap
df.head()

ID  Age  Experience  Income  ZIPCode  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49    91107       4    1.6          1         0   
1   2   45          19      34    90089       3    1.5          1         0   
2   3   39          15      11    94720       1    1.0          1         0   
3   4   35           9     100    94112       1    2.7          2         0   
4   5   35           8      45    91330       4    1.0          2         0   

   Personal_Loan  Securities_Account  CD_Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

<IPython.core.display.Javascript object>

In [12]:
# Noticed above we have an ID column. Let's check to see how many  unique values exists to see if this column will be useful
df.ID.nunique()

5000

<IPython.core.display.Javascript object>

* We have a unique ID for each entry so we can go ahead and drop this column from the dataframe

In [13]:
# dropping ID column since that will not be valueable to our analysis
df.drop(['ID'], axis=1, inplace=True)

<IPython.core.display.Javascript object>

In [14]:
# Updating my list of numerical features to exclude ID
num_cols.pop(0)

'ID'

<IPython.core.display.Javascript object>

In [15]:
# Quick check of datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Age                 5000 non-null   int64  
 1   Experience          5000 non-null   int64  
 2   Income              5000 non-null   int64  
 3   ZIPCode             5000 non-null   int64  
 4   Family              5000 non-null   int64  
 5   CCAvg               5000 non-null   float64
 6   Education           5000 non-null   int64  
 7   Mortgage            5000 non-null   int64  
 8   Personal_Loan       5000 non-null   int64  
 9   Securities_Account  5000 non-null   int64  
 10  CD_Account          5000 non-null   int64  
 11  Online              5000 non-null   int64  
 12  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(12)
memory usage: 507.9 KB


<IPython.core.display.Javascript object>

- Matches with what we saw earlier. All but one category are integars in the dataset. looking at the df sample above, it makes sense but we do have some features that are already encoded or they're zipcodes

In [16]:
# Check the 5 point summary of our categorical features to see if there's anything insightfule from the data
df[cat_cols].describe().T

count       mean          std      min      25%      50%      75%  \
ZIPCode    5000.0  93169.257  1759.455086  90005.0  91911.0  93437.0  94608.0   
Education  5000.0      1.881     0.839869      1.0      1.0      2.0      3.0   

               max  
ZIPCode    96651.0  
Education      3.0

<IPython.core.display.Javascript object>

- Can't take much away from the zipcode data. Will need to dig deeper below
- Education, average customer looks to have more than undergrad education. I'll look deeper into this one too

In [17]:
# Print out the unique values for our categorical features
for col in cat_cols:
    print(f'Unique values in {col}:/n')
    print(f'{df[col].value_counts()}/n')
    print('--'*25)

Unique values in ZIPCode:/n
94720    169
94305    127
95616    116
90095     71
93106     57
        ... 
96145      1
94087      1
91024      1
93077      1
94598      1
Name: ZIPCode, Length: 467, dtype: int64/n
--------------------------------------------------
Unique values in Education:/n
1    2096
3    1501
2    1403
Name: Education, dtype: int64/n
--------------------------------------------------


<IPython.core.display.Javascript object>

- Looking at the zipcode data, it looks like we have a high concentration of customers from the zipcodes 94720, 94305, and 95616. We can visualize this below
- Looking at the education, majority of the customers have an undergrad degree

### Zipcode analysis 

In [18]:
# Importing python library to deal with zipcodes
from uszipcode import SearchEngine
engine = SearchEngine()

<IPython.core.display.Javascript object>

In [24]:
zipcode = df['ZIPCode'].apply(lambda x: engine.by_zipcode(x))

<IPython.core.display.Javascript object>

In [101]:
x = len(df['ZIPCode'])
lat = []
lng = []
for i in range(x):
    try:
        lat[i] = zipcode[i].lat

SyntaxError: unexpected EOF while parsing (3252782720.py, line 6)

ERROR:root:Cannot parse: 7:0: <line number missing in source>
Traceback (most recent call last):
  File "/opt/anaconda3/lib/python3.9/site-packages/lab_black.py", line 218, in format_cell
    formatted_code = _format_code(cell)
  File "/opt/anaconda3/lib/python3.9/site-packages/lab_black.py", line 29, in _format_code
    return format_str(src_contents=code, mode=FileMode())
  File "/opt/anaconda3/lib/python3.9/site-packages/black.py", line 725, in format_str
    src_node = lib2to3_parse(src_contents.lstrip(), mode.target_versions)
  File "/opt/anaconda3/lib/python3.9/site-packages/black.py", line 836, in lib2to3_parse
    raise exc from None
black.InvalidInput: Cannot parse: 7:0: <line number missing in source>


In [77]:
from googlemaps import Client as GoogleMaps
import os
import time

<IPython.core.display.Javascript object>

In [74]:
df[['lat','lng']] = ''

<IPython.core.display.Javascript object>

In [75]:
df.head()

Age  Experience  Income  ZIPCode  Family  CCAvg  Education  Mortgage  \
0   25           1      49    91107       4    1.6          1         0   
1   45          19      34    90089       3    1.5          1         0   
2   39          15      11    94720       1    1.0          1         0   
3   35           9     100    94112       1    2.7          2         0   
4   35           8      45    91330       4    1.0          2         0   

   Personal_Loan  Securities_Account  CD_Account  Online  CreditCard lat lng  
0              0                   1           0       0           0          
1              0                   1           0       0           0          
2              0                   0           0       0           0          
3              0                   0           0       0           0          
4              0                   0           0       0           1

<IPython.core.display.Javascript object>

In [49]:
def extract_latlng(zip):
    api_key = 'AIzaSyBs4qJjfjpS7RK79NxQSic58l7BrXWLZ-U'
    gmaps_client = googlemaps.Client(api_key)

    geocode_result= gmaps_client.geocode(zip)
    return geocode_result[0]['geometry']['location']['lat'], geocode_result[0]['geometry']['location']['lng']

<IPython.core.display.Javascript object>

In [51]:
#df['lat'], df['lng'] = df['ZIPCode'].apply(lambda x: extract_latlng(x))

<IPython.core.display.Javascript object>

In [ ]:
# Extracting lat and long from zipcodes
lat = []
lng = []
               
for i in range(len(df['ZIPCode'])):
    print(i)
    geocode_result= gmaps_client.geocode(df['ZIPCode'][i])
    print(geocode_result[0]['geometry']['location']['lat'])
    print(geocode_result[0]['geometry']['location']['lng'])

In [72]:
gmaps = GoogleMaps('AIzaSyBs4qJjfjpS7RK79NxQSic58l7BrXWLZ-U')

<IPython.core.display.Javascript object>

In [79]:
for x in range(len(df['ZIPCode'])):
    try:
        time.sleep(1) #to add delay in case of large DFs
        geocode_result = gmaps.geocode(df['ZIPCode'][x])
        df['lat'][x] = geocode_result[0]['geometry']['location'] ['lat']
        df['lng'][x] = geocode_result[0]['geometry']['location']['lng']
    except IndexError:
        print("Address was wrong...")
    except Exception as e:
        print("Unexpected error occurred.", e )
df.head()

KeyboardInterrupt: 

<IPython.core.display.Javascript object>

''

<IPython.core.display.Javascript object>

In [ ]:
zipcode.plot(
    kind='scatter', 
    x=zipcode.lat, 
    y=zipcode.lng,
    alpha=0.4, 
    s=zipcode.population/100, 
    label="population", 
    figsize=(10,7),
    c="median_house_value", 
    cmap=plt.get_cmap("jet"), 
    colorbar=True,
)
plt.legend()

## Structure of Numerical features

In [ ]:
# Unique values for each feature
# For each numerical feature compute number of unique entries
unique_values = df.select_dtypes(include='number').nunique().sort_values()
unique_values

In [ ]:
# Five point summary for numerical features
df.describe().T

- Average customer age is 45 years and 50% of the customers are between 35 - 55 years old. 
- Customers have an average of 20 years of experience 
- Average income is a bit more than 73k dollars but 50% of the customers 39-98k dollars
- Zipcode we'll dig into more below
- Family size ranges from 1 to 4
- On average, customers spend 2000 dollars on credit per month
- Education is more categorical so I'll take a deeper look in the univariate analysis to understand the distribution better
- 

In [ ]:
# Plot information
plt.figure(figsize=(20,7))
sns.barplot(x=unique_values.index, y=unique_values.values)
plt.title('Unique values per feature')
plt.show()

## Quality Investigation

Global view of dataset with regards to things like duplicates, missing values and unwatned entries or recording errors

### Duplicates

In [ ]:
# Check for number of duplicates
df.duplicated().sum()

* 